src/server/pps/server/master.go

```diff
@@ -73,28 +75,48 @@ type ppsMaster struct {
 	// the current pps master loses its master status
 	masterCtx context.Context
 
-	// fields for monitorPipeline goros, monitorCrashingPipeline goros, etc.
-	monitorCancelsMu       sync.Mutex
-	monitorCancels         map[string]func() // protected by monitorCancelsMu
-	crashingMonitorCancels map[string]func() // also protected by monitorCancelsMu
+	monitorer *monitorManager
 
 	// fields for the pollPipelines, pollPipelinePods, and watchPipelines goros
 	pollPipelinesMu sync.Mutex
 	pollCancel      func() // protected by pollPipelinesMu
 	pollPodsCancel  func() // protected by pollPipelinesMu
 	watchCancel     func() // protected by pollPipelinesMu
 
+	opsInProcessMu sync.Mutex
+	opsInProcess   map[string]*pipelineOp
+
 	// channel through which pipeline events are passed
 	eventCh chan *pipelineEvent
 }
 
+type monitorManager struct {
```

Comment:
the monitorManager is an attempt to breakup dependencies so that we can pass in the monitor logic to pipelineOp without keeping a back-reference to the API server. I think this helps us know that pipelineOp goros don't do anything unexpected by reaching into the API server

Reply:
Thinking about this a bit more, if we go with the model where the pipelineOp's lifetime matches the pipeline's lifetime, we maybe don't even need `monitorCancels` or `crashingMonitorCancels`--those could maybe just be fields of the op.

In other words, in this PR there are three maps keyed by pipeline:
- `monitorCancels`
- `crashingMonitorCancels`
- `opsInProcess`

I think, instead, we could maybe just have `opsInProcess` (we should still have a central data structure for these: `opManager` or some such). It would map the pipeline name to its op, which would contain `monitorCancel` (if any) and `crashingMonitorCancel` (if any). We might not need `monitorCancelsMu` anymore--if the ppsMaster wants to cancel these goros, it can just cancel the ops via the `opManager`, which has `opsInProcessMu`.

A cool Go trick worth knowing about: if you embed a mutex in a struct, the struct becomes a mutex:
```
type opManager struct {
  sync.Mutex
  ops map[string]*pipelineOp
}
```
then
```
func (m *opManager) bump(pipeline string) {
  m.Lock()  // just lock m itself. If every method starts this way, m is visually thread-safe
  defer m.Unlock()
  op, ok := m.ops[pipeline]
  if ok { op.opChan <- struct{}{}; }
}
```